src/librustc_mir/transform/const_prop.rs

```diff
@@ -575,8 +575,16 @@ impl<'mir, 'tcx> ConstPropagator<'mir, 'tcx> {
             }
 
             // Do not try creating references (#67862)
-            Rvalue::Ref(_, _, place_ref) => {
-                trace!("skipping Ref({:?})", place_ref);
+            Rvalue::AddressOf(_, place) | Rvalue::Ref(_, _, place) => {
+                trace!("skipping AddressOf | Ref for {:?}", place);
+
+                // This may be creating mutable references or immutable references to cells.
+                // If that happens, the pointed to value could be mutated via that reference.
+                // Since we aren't tracking references, the const propagator loses track of what
+                // value the local has right now.
+                // Thus, all locals that have their reference taken
+                // must not take part in propagation.
+                Self::remove_const(&mut self.ecx, place.local);
```

Comment:
I tried simply asserting here, but that won't work for

* zsts or
* locals that got a reference assigned from a constant instead of an `Rvalue::Ref` and are now hitting this `Rvalue::Ref` because the reference is getting reborrowed. This is basically `Projection::Deref` biting us again
* arguments, because they don't get marked as `NoPropagation`